src/main/java/uk/nhs/hee/tis/trainee/forms/service/FormRelocateService.java

```diff
@@ -0,0 +1,186 @@
+import com.amazonaws.xray.spring.aop.XRayEnabled;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import uk.nhs.hee.tis.trainee.forms.dto.enumeration.LifecycleState;
+import uk.nhs.hee.tis.trainee.forms.model.AbstractForm;
+import uk.nhs.hee.tis.trainee.forms.model.FormRPartA;
+import uk.nhs.hee.tis.trainee.forms.model.FormRPartB;
+import uk.nhs.hee.tis.trainee.forms.repository.FormRPartARepository;
+import uk.nhs.hee.tis.trainee.forms.repository.FormRPartBRepository;
+import uk.nhs.hee.tis.trainee.forms.service.exception.ApplicationException;
+
+
+@Slf4j
+@Service
+@Transactional
+@XRayEnabled
+public class FormRelocateService {
+
+  private final FormRPartARepository formRPartARepository;
+  private final FormRPartBRepository formRPartBRepository;
+  protected final AmazonS3 amazonS3;
+
+  protected final ObjectMapper objectMapper;
+
+  @Value("${application.file-store.bucket}")
+  private String bucketName;
+
+  /**
+   * Constructor for Form Relocate service.
+   *
+   * @param formRPartARepository  spring data repository (Form R Part A)
+   * @param formRPartBRepository  spring data repository (Form R Part B)
+   * @param amazonS3              client for S3 service
+   */
+  public FormRelocateService(FormRPartARepository formRPartARepository,
+                             FormRPartBRepository formRPartBRepository,
+                             AmazonS3 amazonS3,
+                             ObjectMapper objectMapper) {
+    this.formRPartARepository = formRPartARepository;
+    this.formRPartBRepository = formRPartBRepository;
+    this.amazonS3 = amazonS3;
+    this.objectMapper = objectMapper;
+  }
+
+  /**
+   * Relocate Form R.
+   */
+  public void relocateFormR(String formId, String targetTrainee) throws IOException {
+
+    // Get Form R from MongoDB by FormId
+    AbstractForm formR = getMoveFormInfoInDb(formId);
+
+    if (formR == null) {
+      log.error("Cannot find FormR with ID " + formId + " from DB.");
+      throw new ApplicationException("Cannot find FormR with ID " + formId + " from DB.");
+    }
+    else {
+      String formType = formR.getFormType();
+      String sourceTrainee = formR.getTraineeTisId();
+      try {
+        updateTargetTraineeInDb(formR, targetTrainee);
+        if (formR.getLifecycleState() != LifecycleState.DRAFT) {
+          relocateFormInS3(formType, formId, sourceTrainee, targetTrainee);
+        }
+      } catch (Exception e) {
+        log.error("Fail to relocate FormR to target trainee: " + e + ". Rolling back...");
+        try {
+          updateTargetTraineeInDb(formR, sourceTrainee);
```

Comment:
It feels like we should be able to use a transaction for this: I don't have any code to offer as a template, but it might be worth looking at the mongo ClientSession to see if that offers a neater way of handling this.

Reply:
Actually, the more I look at this, the more it seems like it is actually fine: spanning a transaction over a DB and an AWS action might be a bit of a stretch!

Reply:
I was struggling to implement roll back in a tidy way as it needs to perform across DB and S3 operations. this is the version I have tried my best to make it simple.
I tried to use `@transactional` annotation, but it doesn't seem to work on these internal method calls within the same class. Any ideas about rolling back are welcomed 👂